docs/design/password-key-derivation.md

```diff
@@ -0,0 +1,76 @@
+
+## Design
+
+Uplink always stores derived keys, never the input material.
+
+The satellite will add a new endpoint, `GetSaltEntropy` that uses the api key to find the project, and optionally takes a bucket name. It is important that the result is stable, even if the bucket is renamed. For example the result of `GetSaltEntropy("bucket1")` should be the same as `GetSaltEntropy("bucket2")` even after `"bucket1"` is renamed to `"bucket2"`. Note that there are still significant challenges to bucket renaming, and we don't support it right now.
+
+When deriving a root key from an api key, password, optional bucket name and optional encrypted path, the following algorithm will be used
+
+```
+saltEntropy = GetSaltEntropy(bucketName or None)
+mixedSalt   = hmac(hash:sha256, secret:password, data:saltEntropy)
+pathSalt    = hmac(hash:sha256, secret:mixedSalt, data:encryptedPath or "")
+rootKey     = argon2id(salt:salt, password:password)
+```
+
+In the terminology of section 4.11 in the [whitepaper](https://storj.io/storjv3.pdf), the algorithm above describes how to derive s<sub>0</sub>. Subsequent secrets are derived using the unencrypted path components and will end up encrypted as described in the whitepaper and appended to the base encrypted path used when deriving s<sub>0</sub>.
+
+The network request to `GetSaltEntropy` only has to occur when a user is inputting a password. Hypothetically, this may happen when
+
+1. Doing initial setup and creating a default password for buckets.
+2. Creating a bucket with different encryption than the default password.
+3. Importing a share that either does not contain an encryption key or is flagged to require one.
+4. Exporting a share for another to use and a different password is desired.
+
+Importantly, it does not happen when interacting with files in the bucket: the key should already be derived and stored in some manner.
+
+## Rationale
+
+This design accomplishes all of the requirements listed above.
+
+- Each root key requires some secret that necessarily comes from the user in some way.
+- The algorithm is deterministic, so the user gets the same keys on any machine where she inputs the same secrets.
+- Root keys for a sub-path in a bucket are accomplished by providing a non-empty encrypted path when deriving the root key.
+- Root keys do not depend on the name of the bucket they are in: just the salt entropy associated with that bucket, allowing renames.
+- Because the entropy is added as the salt in the `argon2id` step, dictionary attacks on low-entropy passwords cannot be precomputed.
+- The default password can be used for any bucket and will return a distinct key as long as the salt entropy differs.
+
+Some other points of consideration include
+
+- Clients can ignore the salt entropy if they wish and know they have a strong password. Indeed, they can ignore all of this and do their own key derivation.
```

Comment:
Can we specify what level of library support we intend to provide for such a use case? Would they have to reimplement large parts of libuplink or will there be a way to just supply your own `rootKey` directly? I think that would probably be the route to take for people who already have high-entropy keys (a separate keyfile, a key from a hardware token, etc).

Reply:
I don't think we need to have great support or anything, but a good goal would be to allow people to just input whatever bytes they want into their configuration entries for the keys and have it work.

Those details of course depend on how the keys are stored exactly, but I think we should be storing them such that minimal transformations are applied to them before use.

Reply:
Please to add a statement to that effect (at the minimum, we will allow the user to specify an already-formed encryption key if they don't want one derived from a passphrase)

Reply:
:+1: